src/base/castlerandom.pas

```diff
@@ -67,22 +68,50 @@ procedure TCastleRandom.Initialize(RandomSeed: LongWord);
 begin
   if RandomSeed = 0 then
   begin
+    Seed := GetRandomSeed;
+  end
+  else seed := LongInt(RandomSeed);
+end;
+
+{$IFDEF Linux}
+function DEV_URANDOM: longint;
+var dev_rnd: file of integer;
+begin
+  { algorithm according to http://wiki.freepascal.org/Dev_random
+   /dev/urandom is a native *nix very high-quality random number generator.
+   it's 1000 times slower than CastleRandom,
+   but provides a perfect seed initialization. }
+  //filemode := 0;
+  AssignFile(dev_rnd, '/dev/urandom');
+  reset(dev_rnd);
+  read(dev_rnd,result);
+  CloseFile(dev_rnd);
+end;
+{$ENDIF}
+
```

Comment:
I've fixed randomization on Windows machines. See https://github.com/eugeneloza/castle-engine/commit/9cc21674b8345c0a153f8e40aab9eeb452fdec88 and https://github.com/eugeneloza/castle-engine/commit/c80b92f04245e6d7e13bb52f8763d90ae499cc6c

Not as beautiful as /dev/urandom thou.
 I've tested it thoroughly and it provides really unpredictable random sequence
 by using a 64-bit xorshift algorithm. It also solves problem of "less than 15ms
 randomization".
Tests show it's working as expected https://github.com/eugeneloza/castle-engine/commit/2658d669fce86c18b727365d244f0e0448f0b0b3

P.S. I could never think SysUtils would have such an ugly random seed
 initialization algorithm :). They do so much effort to make the
 random sequence nice, but they don't do anything to have really different
 random sequences... And some users even argue that "don't run randomize more
 than once in your program!", etc.

> And if there is a problem in FPC implementation of Randomize, then we should fix the FPC too

Yes, maybe it'd be ok to submit this procedure as a patch to SysUtils because it's much more random and looks absolutely thread-safe.

> I'm unsure about having a special implementation for Linux.

I've introduced and commented out {$DEFINE USE_DEV_URANDOM}. So initialization works according to equal algorithm on any system at the moment. As far as I see, dev/urandom is not implemented on all unix-like systems, e.g. OpenBSD doesn't have it. However, it looks like all systems have /dev/random implemented (which is slower in Linux). Moreover, some systems might require additional previleges from user to read /dev/random.
So, basically, I'm not sure if it's a good idea to initialize it anywhere except for Linux machines.
The new 64-bit algorithm I've made looks to be very nice as it grabs and integrates 3 semi-independent semi-random variables as CastleTimeUtils.gettickcount64, SysUtils.now and memory allocation address of a local variable. Of course, /dev/urandom is better, but we're not pursing cryptographic safety, but just need to make sure 2 random sequences cannot be equal otherwise than by a random coincidence with a chance of 1/max(longword).